[view-transitions] Captured root element with margins display snapshots in incorrect position.

mattwoodrow opened the pull request:
#### 2058eceb3fd5f84b7d394aa7281110088f9f62e1
<pre>
[view-transitions] Captured root element with margins display snapshots in incorrect position.
<a href="https://bugs.webkit.org/show_bug.cgi?id=274397">https://bugs.webkit.org/show_bug.cgi?id=274397</a>
&lt;<a href="https://rdar.apple.com/128405732">rdar://128405732</a>&gt;

Reviewed by Tim Nguyen.

If the root element is captured in the view transition, then we effectively do snapshotting
of the RenderView instead.

Computing the element-to-screen transform should be skipped, since it would include any
margin on the root element relative the view.

Most of the code in RenderLayerBacking that prepares a captured layer for attachment
into the pseduo element tree also needs to operate on the RenderView&apos;s layer in this case,
so a new &apos;effectiveCapturedInViewTransition&apos; added to make this simpler.

* LayoutTests/TestExpectations:
* Source/WebCore/dom/ViewTransition.cpp:
(WebCore::ViewTransition::copyElementBaseProperties):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::paintList):
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateTransform):
(WebCore::RenderLayerBacking::updateChildrenTransformAndAnchorPoint):
(WebCore::RenderLayerBacking::shouldClipCompositedBounds const):
(WebCore::RenderLayerBacking::updateGeometry):
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::requiresCompositingForViewTransition const):
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::effectiveCapturedInViewTransition const):
* Source/WebCore/rendering/RenderObject.h:

Canonical link: <a href="https://commits.webkit.org/279372@main">https://commits.webkit.org/279372@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9adaab504eb716f872fd46614248377b9594f92e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/53312 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/32663 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/5810 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/56591 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/4038 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/55618 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/39999 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/3788 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/43226 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/2641 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/55410 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/30783 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/46045 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/24356 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/27694 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/3365 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/2194 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/49455 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/3529 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/58187 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/28461 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/3543 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/50622 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/29676 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/46254 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/49953 "Found 1 new API test failure: /WebKitGTK/TestUIClient:/webkit/WebKitWebView/display-usermedia-permission-request (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11617 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/30600 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/29439 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m reviewed (approved):
Would be nice to have a new WPT, but optional given the progression

nt1m commented on Source/WebCore/page/LocalFrameView.cpp:
```diff
@@ -3343,7 +3343,7 @@ void LocalFrameView::setNeedsLayoutAfterViewConfigurationChange()
 {
     RenderView* renderView = this->renderView();
-    if (renderView && renderView->usesCompositing()) {
+    if (renderView && renderView && renderView->usesCompositing()) {
```
```suggestion
    if (renderView && renderView->usesCompositing()) {
```
nit: double `renderView &&`

also, line above could use `auto*`

nt1m commented on Source/WebCore/page/LocalFrameView.cpp:
```diff
@@ -3353,7 +3353,7 @@ void LocalFrameView::setNeedsCompositingConfigurationUpdate()
 {
     RenderView* renderView = this->renderView();
-    if (renderView->usesCompositing()) {
+    if (renderView && renderView->usesCompositing()) {
```
ditto regarding auto

nt1m commented on Source/WebCore/page/LocalFrameView.cpp:
```diff
@@ -3363,7 +3363,7 @@ void LocalFrameView::setNeedsCompositingGeometryUpdate()
 {
     RenderView* renderView = this->renderView();
-    if (renderView->usesCompositing()) {
+    if (renderView && renderView->usesCompositing()) {
```
ditto

nt1m commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -10527,6 +10527,7 @@ bool Document::activeViewTransitionCapturedDocumentElement() const
 void Document::setActiveViewTransition(RefPtr<ViewTransition>&& viewTransition)
 {
     m_activeViewTransition = WTFMove(viewTransition);
+
```
nit: blank extra line

mattwoodrow committed:
[view-transitions] Captured root element with margins display snapshots in incorrect position.
https://bugs.webkit.org/show_bug.cgi?id=274397
<rdar://128405732>

Reviewed by Tim Nguyen.

If the root element is captured in the view transition, then we effectively do snapshotting
of the RenderView instead.

Computing the element-to-screen transform should be skipped, since it would include any
margin on the root element relative the view.

Most of the code in RenderLayerBacking that prepares a captured layer for attachment
into the pseduo element tree also needs to operate on the RenderView's layer in this case,
so a new 'effectiveCapturedInViewTransition' added to make this simpler.

* LayoutTests/TestExpectations:
* Source/WebCore/dom/ViewTransition.cpp:
(WebCore::ViewTransition::copyElementBaseProperties):
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::paintList):
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateTransform):
(WebCore::RenderLayerBacking::updateChildrenTransformAndAnchorPoint):
(WebCore::RenderLayerBacking::shouldClipCompositedBounds const):
(WebCore::RenderLayerBacking::updateGeometry):
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::requiresCompositingForViewTransition const):
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::effectiveCapturedInViewTransition const):
* Source/WebCore/rendering/RenderObject.h:

Canonical link: https://commits.webkit.org/279372@main
